# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

### 導入數據分極需要的庫，全通過pandas的 read_csv函數，將原始數據文件"e_commerce.csv"里的數據內，解析為DataFrame, 並賦值給變量original_data

In [1]:
import pandas as pd

In [2]:
original_data = pd.read_csv("e_commerce.csv")

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

In [4]:
original_data.sample(10)

InvoiceNo StockCode                     Description  Quantity  \
151062    549537     22257          FELT FARM ANIMAL SHEEP         1   
412512    572291     22335  HEART DECORATION PAINTED ZINC          6   
80488     543040    79191B        RETRO PLASTIC POLKA TRAY         2   
158876    550318     21790              VINTAGE SNAP CARDS         3   
110099    545667    47593B           SCOTTIE DOGS BABY BIB         1   
353412    567802     84879   ASSORTED COLOUR BIRD ORNAMENT        16   
65542     541696     21363         HOME SMALL WOOD LETTERS         1   
471009    576618     23582    VINTAGE DOILY JUMBO BAG RED          1   
287026    562046     21668  RED STRIPE CERAMIC DRAWER KNOB        12   
145286    548893    84596F    SMALL MARSHMALLOWS PINK BOWL         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
151062    4/8/2011 17:22       1.25         NaN  United Kingdom  
412512  10/23/2011 12:46       0.65     16931.0  United Kingdom  
80488     2/2/2011 17:06       0.85     17337.0  United Kingdom  
158876   4/17/2011 12:25       0.85     16085.0  United Kingdom  
110099    3/4/2011 16:07       0.83         NaN  United Kingdom  
353412   9/22/2011 12:06       1.69     14299.0  United Kingdom  
65542    1/20/2011 18:08      10.79         NaN  United Kingdom  
471009  11/15/2011 17:00       4.13         NaN  United Kingdom  
287026    8/2/2011 10:34       1.25     12378.0     Switzerland  
145286    4/4/2011 15:54       0.83         NaN  United Kingdom

從抽樣的10行數據來看，數據符合"每列是一個變量，每個單元格是一個變量，每個單元格是一個值"， 具體來看每行是關於某商品的一次交易，每列是交易相關的各個變量，因此不存在結構性問題。

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


從輸出結果看，共有541909條曲察值，而Decription、CustomerID變量存在缺失值。
此外，InvoiceDate的數據類型應為日期，Customer的數據類型應該為字符串，應當進行數據轉換。

#### 評估缺失數據

在了解Description存在缺失值後，根據條件提取出缺失觀察值。

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
535322    581199     84581         NaN        -2  12/7/2011 18:26        0.0   
535326    581203     23406         NaN        15  12/7/2011 18:31        0.0   
535332    581209     21620         NaN         6  12/7/2011 18:35        0.0   
536981    581234     72817         NaN        27  12/8/2011 10:33        0.0   
538554    581408     85175         NaN        20  12/8/2011 14:06        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
535322         NaN  United Kingdom  
535326         NaN  United Kingdom  
535332         NaN  United Kingdom  
536981         NaN  United Kingdom  
538554         NaN  United Kingdom  

[1454 rows x 8 columns]

從輸出結果看，這些缺失Description的交易數據，UnitPrice都為0，為了驗證猜想，我們增加篩選條件，看是否存在Description變量缺失且UnitPrince不為0的數據。

In [7]:
original_data[(original_data["Description"].isnull()) & (original_data["UnitPrice"] != 0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

篩選出來結果為0條，說明缺失Description值的數據，同時也不具備有效的UnitPrice值。
Description為產品名稱，UnitPrice為產品單價，都是進行後續商品交易分析的重要變量。，如果它們同時缺失，我們認為數據無法提供有效含義，因此這些後續可以被刪除。

In [8]:
original_data[(original_data["CustomerID"].isnull())]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
541536    581498    85099B          JUMBO BAG RED RETROSPOT         5   
541537    581498    85099C   JUMBO  BAG BAROQUE BLACK WHITE         4   
541538    581498     85150    LADIES & GENTLEMEN METAL SIGN         1   
541539    581498     85174                S/4 CACTI CANDLES         1   
541540    581498       DOT                   DOTCOM POSTAGE         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
541536  12/9/2011 10:26       4.13         NaN  United Kingdom  
541537  12/9/2011 10:26       4.13         NaN  United Kingdom  
541538  12/9/2011 10:26       4.96         NaN  United Kingdom  
541539  12/9/2011 10:26      10.79         NaN  United Kingdom  
541540  12/9/2011 10:26    1714.17         NaN  United Kingdom  

[135080 rows x 8 columns]

CustomerID表示客戶編號，不是分析暢銷商品的必要變量。並且從輸出結果來看，有些CustomerID的交易數據仍然有效
，因此此變量為空的觀察值。

#### 評估重複數據

根據變量的含義來看，雖然InvoiceNo、StockCode和CustomerID都是唯一標識符，但一次交易可能包含多件商品，因此InvoiceNo可以重複，不同交易可以包含同件商品，因此StockCode可以存在重複，顧客可以進行多次交易或下單多個商品，因此CustomerID也可以存在重複。

那麼針對此數據集，我們無需評估重複數據。

#### 評估不一致數據

不一致數據可能存在"Country"變量中，我們要查看是否存在多個不同值指代同一個國家的情況。

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          495266
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
China                      288
Singapore                  229
USA                        218
UK                         211
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United States               73
United Arab Emirates        68


從country變量來看，"USA"、"United States"均在表示美國，"United Kindom"、"UK"、"U.K."均在表示英國，因此應該在清洗步對這些值進行統一，只保留一個指代值

#### 評估無效或錯誤數據

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

從輸出結果來看，Quantity和UnitPrice存在負數，會對後續數值產生影響。
因此，我們要先篩選出quantity為負數的觀察值，進一步評估其含義

In [11]:
original_data[original_data["Quantity"] < 0]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[10624 rows x 8 columns]

從結果來看Quantity變數為負數的觀察值，InvoiceNO似乎都以C開頭，表示訂單被取消。
但此猜測不一定正確，為了驗證猜想，我們增加篩選條件，看是否存在Quantity為負數且InvoiceNO開頭不為C的觀察值

In [12]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C")]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777         NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C         NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712         NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028         NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067         NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...         ...       ...              ...        ...   
535333    581210     23395       check       -26  12/7/2011 18:36        0.0   
535335    581212     22578        lost     -1050  12/7/2011 18:38        0.0   
535336    581213     22576       check       -30  12/7/2011 18:38        0.0   
536908    581226     23090     missing      -338   12/8/2011 9:56        0.0   
538919    581422     23169     smashed      -235  12/8/2011 15:24        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
535333         NaN  United Kingdom  
535335         NaN  United Kingdom  
535336         NaN  United Kingdom  
536908         NaN  United Kingdom  
538919         NaN  United Kingdom  

[1336 rows x 8 columns]

以上猜想被證實錯誤，因為還存在在Quantity為負數且InvoiceNO開頭不為C的觀察值。
但根據結果，這些篩選出的觀察值的UnitPrice的觀察值均為0，因此增加UnitPrice的條件進行驗證。

In [13]:
original_data[(original_data["Quantity"] < 0) & (original_data["InvoiceNo"].str[0] != "C") & (original_data["UnitPrice"] != 0 )]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根據輸出結果，猜想得到驗證，當Quantity為負數時，觀察值滿足以下條件之一:
InvoiceNo以C開頭，表示訂單取消
UnitPrice為0，說明商品單價為0英磅
這些交易數據不是有效成交數據，不貢獻銷售，不在後續分析範圍內，因此我們將在數據清理步驟將Quantity為負數的觀察值刪除

In [14]:
original_data[(original_data["UnitPrice"] < 0 )]

InvoiceNo StockCode      Description  Quantity      InvoiceDate  \
299983   A563186         B  Adjust bad debt         1  8/12/2011 14:51   
299984   A563187         B  Adjust bad debt         1  8/12/2011 14:52   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

從輸出結果看，UnitPrice為負數的觀察值都是壞帳調賬，不屬于實際商品交易數據，因此也在數據清理步驟中也將其刪除。

## 清理数据

数据每列的含义如下：

    InvoiceNo: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
    StockCode: 产品代码。5位数，作为产品的唯一标识符。
    Description: 产品名称。
    Quantity: 产品在交易中的数量。
    InvoiceDate: 发票日期和时间。交易发生的日期和时间。
    UnitPrice: 单价。价格单位为英镑（£）。
    CustomerID: 客户编号。5位数，作为客户的唯一标识符。
    Country: 国家名称。客户所居住的国家的名称。



為了后分開清理的和原的粓據，我們將創建新的變量clearned_data,讓𩫷為original_data複製出的副本，
我們之後的清理步驟都將被運用在cleaned_data上



In [15]:
cleared_data = original_data.copy()
cleared_data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom  
...                 ...        ...         ...             ...  
541904  12/9/2011 12:50       0.85     12680.0          France  
541905  12/9/2011 12:50       2.10     12680.0          France  
541906  12/9/2011 12:50       4.15     12680.0          France  
541907  12/9/2011 12:50       4.15     12680.0          France  
541908  12/9/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

把InvoiceNDate變量的數據類型轉換成日期時間類型

In [16]:
cleared_data["InvoiceDate"] = pd.to_datetime(cleared_data["InvoiceDate"])
cleared_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

把CustomerID轉換成字符串

In [17]:
cleared_data["CustomerID"] = cleared_data["CustomerID"].astype(str)
cleared_data["CustomerID"]

0         17850.0
1         17850.0
2         17850.0
3         17850.0
4         17850.0
           ...   
541904    12680.0
541905    12680.0
541906    12680.0
541907    12680.0
541908    12680.0
Name: CustomerID, Length: 541909, dtype: object

把.0刪除

In [18]:
cleared_data["CustomerID"] = cleared_data["CustomerID"].str.slice(0, -2)
cleared_data["CustomerID"]

0         17850
1         17850
2         17850
3         17850
4         17850
          ...  
541904    12680
541905    12680
541906    12680
541907    12680
541908    12680
Name: CustomerID, Length: 541909, dtype: object

把Description缺失的觀察值刪除，並查看刪除後該列空缺值和:

In [21]:
cleared_data.dropna(subset=["Description"], inplace = True)

In [22]:
cleared_data["Description"].isnull().sum()

0

把Country變量值"USA"替換為"United State", 並檢察替換"USA"變量值個數:

In [23]:
cleared_data["Country"] = cleared_data["Country"].replace({"USA":"United States"})

In [24]:
len(cleared_data[cleared_data["Country"] == "USA"])

0

把Country變量值"UK", "U.K."替換為"United Kindom", 並檢察替換"UK", "U.K."變量值個數:

In [25]:
cleared_data["Country"] = cleared_data["Country"].replace({"UK":"United Kingdom","U.K.":"United Kingdom"})

In [27]:
len(cleared_data[cleared_data["Country"] == "UK"])
len(cleared_data[cleared_data["Country"] == "U.K."])

0

把Quantity缺失的觀察值刪除，並查看刪除後該列空缺值和:

In [30]:
cleared_data = cleared_data[cleared_data["Quantity"] >= 0 ]

In [31]:
len(cleared_data[cleared_data["Quantity"] < 0])

0

In [ ]:
把UnitPrice缺失的觀察值刪除，並查看刪除後該列空缺值和:

In [33]:
cleared_data = cleared_data[cleared_data["UnitPrice"] >= 0 ]

In [34]:
len(cleared_data[cleared_data["UnitPrice"] < 0])

0

## 保存清理后的数据

In [35]:
cleaned_data = cleared_data

In [ ]:
完成數據清洗後，把乾淨的數據保存到新的文件夾裡，文件名為e_commerce_cleaned.csv

In [38]:
cleaned_data.to_csv("e_commerce_cleaned.csv", index = False)

In [39]:
pd.read_csv("e_commerce_cleaned.csv").head()

C:\Users\Tyler\AppData\Local\Temp\ipykernel_568\2686214642.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  pd.read_csv("e_commerce_cleaned.csv").head()


InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice CustomerID         Country  
0  2010-12-01 08:26:00       2.55      17850  United Kingdom  
1  2010-12-01 08:26:00       3.39      17850  United Kingdom  
2  2010-12-01 08:26:00       2.75      17850  United Kingdom  
3  2010-12-01 08:26:00       3.39      17850  United Kingdom  
4  2010-12-01 08:26:00       3.39      17850  United Kingdom